keps/sig-cluster-lifecycle/0015-kubeadm-join-master.md

```diff
@@ -0,0 +1,431 @@
+
+4. In case of `join --master` [New step]
+
+   1. Apply master taint and label to the node.
+
+      > This action is executed using the admin.conf identity created above;
+
+   2 Update the `kubeadm-config` configMap with the information about the new master node
+
+#### dynamic workflow (advertise-address == `controlplaneAddress`)
+
+There are many ways to configure an highly available cluster.
+
+Among them, the approach best suited for a dynamic bootstrap workflow requires the
+user to set the `--apiserver-advertise-address` of each master, including the bootstrap master
+itself, equal to the `controlplaneAddress` endpoint provided during kubeadm init
+(the IP/DNS of the external load balancer).
+
+By using the same advertise address for all the IP masters, `kubeadm init` can create
+a unique API server serving certificate that could be shared across many masters nodes;
+no changes will be required to this certificate when adding/removing master nodes.
+
+Please note that:
+
+- if the user is not planning to distribute the apiserver serving certificate among masters,
+  kubeadm will generate a new apiserver serving certificate “almost equal” to the certificate
+  created on the bootstrap master (it differs only for the domain name of the joining master)
+
+#### Static workflow (advertise-address != `controlplaneAddress`)
+
+In case of a static bootstrap workflow the final layout of the controlplane - the number, the
+name and the IP of master nodes - is know in advance.
+
+Given such information, the user can choose a different approach where each master has a
+specific apiserver advertise address different from the `controlplaneAddress`.
+
+Please note that:
+
+- if the user is not planning to distribute the apiserver certificate among masters, kubeadm
+  will generate a new apiserver serving certificate with the required SANS
+- if the user is planning to distribute the apiserver certificate among masters, he/she/the
+  operator is required to provide during `kubeadm init` the list of masters/the list of IP
+  addresses for all the masters as alternative names for the API servers certificate, thus
+  allowing the proper functioning of all the API server instances that will join
+
+#### Strategies for deploying control plane components
+
+As of today kubeadm supports two solutions for deploying control plane components:
+
+1. Control plane deployed as static pods (current kubeadm default)
+2. Self-hosted control plane in case of `--feature-gates=SelfHosting=true`
+
+The proposed solution for case 1. "Control plane deployed as static pods", assumes
+that the `kubeadm join --master` flow will take care of creating required kubeconfig
+files and required static pod manifests.
+
+Self-hosted control plane are not supported in this release of the proposal
+(`kubeadm join --master` will fail immediately), but nothing in this proposal should
+prevent to address this in subsequent phases.
+
+#### Strategies for distributing cluster certificates
```

Comment:
Please dive into how the CSR API could be utilized in an external CA-world.
It'd be cool to just use the CSR API, but we're stuck needing the front-proxy clientcert for the API server and the SA key that is not bound to the CA.

Reply:
I would prefer to keep this out of scope of this scope, because it does not apply specifically to the multi-master scenario (e.g. it is relevant also for the existing `kubeadm join` workflow).

Never the less I totally agree the we should document how the CSR API could be used in an external CA-world, but I don't have real expertise in this. Eventually, let's raise the topic at the sig-meeting